server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,439 @@
+        generateScheduledEventsFrom (stsMaxScheduledTime stats) sti
+  case scheduledEvents of
+    []     -> pure ()
+    events -> do
+      let insertScheduledEventsSql = TB.run $ toSQL
+            SQLInsert
+              { siTable    = scheduledEventsTable
+              , siCols     = map unsafePGCol ["name", "scheduled_time"]
+              , siValues   = ValuesExp $ map (toTupleExp . toArr) events
+              , siConflict = Just $ DoNothing Nothing
+              , siRet      = Nothing
+              }
+      Q.unitQE defaultTxErrorHandler (Q.fromText insertScheduledEventsSql) () False
+  where
+    toArr (ScheduledEventSeed n t) = [(triggerNameToTxt n), (formatTime' t)]
+    toTupleExp = TupleExp . map SELit
+
+insertScheduledEvents :: [ScheduledEventSeed] -> Q.TxE QErr ()
+insertScheduledEvents events = do
+  let insertScheduledEventsSql = TB.run $ toSQL
+        SQLInsert
+          { siTable    = scheduledEventsTable
+          , siCols     = map unsafePGCol ["name", "scheduled_time"]
+          , siValues   = ValuesExp $ map (toTupleExp . toArr) events
+          , siConflict = Just $ DoNothing Nothing
+          , siRet      = Nothing
+          }
+  Q.unitQE defaultTxErrorHandler (Q.fromText insertScheduledEventsSql) () False
+  where
+    toArr (ScheduledEventSeed n t) = [(triggerNameToTxt n), (formatTime' t)]
+    toTupleExp = TupleExp . map SELit
+
+generateScheduledEventsFrom :: UTCTime -> ScheduledTriggerInfo-> [ScheduledEventSeed]
+generateScheduledEventsFrom startTime ScheduledTriggerInfo{..} =
+  let events =
+        case stiSchedule of
+          AdHoc _ -> empty -- ad-hoc scheduled events are created through 'create_scheduled_event' API
+          Cron cron -> generateScheduleTimes startTime 100 cron -- by default, generate next 100 events
+   in map (ScheduledEventSeed stiName) events
+
+-- | Generates next @n events starting @from according to 'CronSchedule'
+generateScheduleTimes :: UTCTime -> Int -> CronSchedule -> [UTCTime]
+generateScheduleTimes from n cron = take n $ go from
+  where
+    go = unfoldr (fmap dup . nextMatch cron)
+
+-- | Generates events @(from, till]@ according to 'CronSchedule'
+-- generateScheduleTimesFrom :: UTCTime -> UTCTime -> CronSchedule -> [UTCTime]
+-- generateScheduleTimesFrom from till cron = takeWhile (<= till) $ go from
+--   where
+--     go = unfoldr (fmap dup . nextMatch cron)
+
+processScheduledQueue
+  :: HasVersion
+  => L.Logger L.Hasura
+  -> LogEnvHeaders
+  -> HTTP.Manager
+  -> Q.PGPool
+  -> IO SchemaCache
+  -> IO ()
+processScheduledQueue logger logEnv httpMgr pgpool getSC =
+  forever $ do
+    scheduledTriggersInfo <- scScheduledTriggers <$> getSC
+    scheduledEventsE <-
+      runExceptT $
+      Q.runTx pgpool (Q.ReadCommitted, Just Q.ReadWrite) getScheduledEvents
+    case scheduledEventsE of
+      Right partialEvents ->
+        sequence_ $
+        flip map partialEvents $ \(ScheduledEventPartial id' name st payload tries)-> do
```

Comment:
`sequence_ $ flip map partialEvents $ \x -> ...` can be shortened to `for_ partialEvents $ \x -> ...`.